### PR TITLE
Support Subscribe Announces 

### PR DESCRIPTION
**Pull request by suhasHere**

### CODE REVIEW

File: src/client.cpp [hunk 1]
```diff
@@ -299,6 +299,37 @@ namespace quicr {
...
+                }
+                return true;
+            }
+            case messages::MoqMessageType::ANNOUNCE: {
+                messages::MoqAnnounce msg;
+                msg_bytes >> msg;
+
+                SPDLOG_LOGGER_DEBUG(logger_, "Received announce on conn_id: {0}", conn_ctx.connection_handle);
+
+                for (const auto& entry : conn_ctx.announce_subscriptions) {
```
Comment (GhostofCookie) [inline src/client.cpp:322]: Prefer structured bindings to avoid using hard to read `first` and `second`.

File: src/messages.cpp [hunk 2]
```diff
@@ -675,6 +675,110 @@ namespace quicr::messages {
         return buffer >> msg.new_session_uri;
     }
 
+    //
+    // Subscribe Announces
+    //
+
+    //
+    // Announce
+    //
```
Comment (GhostofCookie) [inline src/messages.cpp:684]: Double comment section

File: src/server.cpp [hunk 3]
```diff
@@ -29,6 +29,12 @@ namespace quicr {
 
     void Server::ResolveAnnounce(ConnectionHandle, const TrackNamespace&, const AnnounceResponse&) {}
 
+    void Server::SubscribeAnnouncesReceived(ConnectionHandle connection_handle, TrackNamespace& track_namespace_prefix)
```
Comment (GhostofCookie) [inline src/server.cpp:32]: Remove unused names to avoid compiler complaining.

File: src/transport.cpp [hunk 4]
```diff
@@ -490,6 +520,47 @@ namespace quicr {
...
+            return;
+        }
+
+        auto it = conn_it->second.announce_subscriptions.find(ns_prefix);
+        if (it != conn_it->second.announce_subscriptions.end()) {
+            // do nothing
+            return;
+        }
+
+        conn_it->second.announce_subscriptions[ns_prefix] = std::move(handler);
```
Comment (GhostofCookie) [inline src/transport.cpp:545]: `it` already is `conn_it->second.announce_subscriptions[ns_prefix]`, we don't need to look it up again. In map this is effectively an O(1) operation though, so maybe it doesn't matter.

File: include/quicr/detail/messages.h [hunk 5]
```diff
@@ -59,6 +59,10 @@ namespace quicr::messages {
 
         GOAWAY = 0x10,
 
+        SUBSCRIBE_ANNOUNCES = 0x11,
```
Comment (TimEvens) [inline include/quicr/detail/messages.h:62]: All of these need to be style formatted. For example `kSubscribeAnnounces`

File: include/quicr/subscribe_announces_handler.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        TrackNamespace GetTrackNamespacePrefix() const noexcept { return track_namespace_prefix_; }
+
+        // --------------------------------------------------------------------------
+        // Public Virtual API callback event methods
+        // --------------------------------------------------------------------------
+        /** @name Callbacks
+         */
+        ///@{
+
+        virtual void MatchingTrackNamespaceReceived([[maybe_unused]] const quicr::TrackNamespace& track_namespace) {}
```
Comment (TimEvens) [inline include/quicr/subscribe_announces_handler.h:57]: name change

File: src/client.cpp [hunk 7]
```diff
@@ -302,6 +302,37 @@ namespace quicr {
                 conn_ctx.setup_complete = true;
                 return true;
             }
+            case messages::ControlMessageType::SUBSCRIBE_ANNOUNCES_OK: {
```
Comment (TimEvens) [inline src/client.cpp:305]: `kSubscribeAnnouncesOk`

File: src/client.cpp [hunk 8]
```diff
@@ -302,6 +302,37 @@ namespace quicr {
...
+                  logger_, "Received subscribe_announce ok, conn_id: {0}", conn_ctx.connection_handle);
+
+                auto it = conn_ctx.announce_subscriptions.find(msg.track_namespace_prefix);
+                if (it->second.get()->GetStatus() != SubscribeAnnouncesHandler::Status::kOk) {
+                    // TODO : report status to the handler
+                    it->second.get()->SetStatus(SubscribeAnnouncesHandler::Status::kOk);
+                }
+                return true;
+            }
+            case messages::ControlMessageType::ANNOUNCE: {
```
Comment (TimEvens) [inline src/client.cpp:319]: `kAnnounce`

File: src/server.cpp [hunk 9]
```diff
@@ -111,10 +118,6 @@ namespace quicr {
 
                 // TODO(tievens): add filter type when caching supports it
                 SubscribeReceived(conn_ctx.connection_handle, msg.subscribe_id, msg.track_alias, tfn, {});
-
-                // TODO(tievens): Delay the subscribe OK till ResolveSubscribe() is called
-                SendSubscribeOk(conn_ctx, msg.subscribe_id, kSubscribeExpires, false);
```
Comment (TimEvens) [inline src/server.cpp:116]: Removing this would require it to be added to resolve subscribe, right?  I don't see that changed added.

File: cmd/examples/server.cpp [hunk 10]
```diff
@@ -86,6 +93,15 @@ namespace qserver_vars {
     std::unordered_map<quicr::messages::TrackAlias,
                        std::unordered_map<quicr::ConnectionHandle, std::shared_ptr<quicr::SubscribeTrackHandler>>>
       pub_subscribes;
+
+    static std::string to_ascii(const std::vector<Span<const uint8_t>>& data)
+    {
+        std::stringstream output;
+        for (const auto& entry : data) {
+            output << entry.data() << ",";
+        }
```
[the remark this answers is not in the document]
Comment (GhostofCookie) [reply cmd/examples/server.cpp:102]: But what is the format? Commas? Slashes? Colons? Does it matter and we can just pick one? Wasn't specified, so wasn't added.

File: include/quicr/detail/messages.h (hunk 5 above)
Comment (suhasHere) [reply include/quicr/detail/messages.h:62]: yes I saw the incosistency. I feel that can be done in a follow up PR.

File: src/server.cpp (hunk 9 above)
Comment (suhasHere) [reply src/server.cpp:116]: good catch . That was error . I will add it back

**Review by TimEvens — COMMENTED** [review]
Looks good, but doesn't seem to be working correctly... likely because `Contains()` is being used to match the track namespace prefix.  

For example, subscribe to announces `a,b`... Publish announces of `a,b` works as expected, but does not work for `a,b,?` and `a,b,**`


There isn't a debug message for the announce subscribe, but I did announce subscribe to `x,y` and then published announce using `x,y,z` which resulted in the below log message.

```
[2024-11-01 15:18:59.277] [info] [server.cpp:349] No match Found for announced namespace. prefix: x/y, namespace: x/y
```

Using anything other than `x,y` for the announce, it doesn't match.  If I publish announce using `x,y` it does work with:

```
[2024-11-01 15:22:24.285] [info] [server.cpp:342] Found a match for announced namespace. prefix: x/y, namespace: x/y
```

So in other words, prefix matching is not working for announce subscribes.

**Comment by TimEvens** [comment]
Many changes since this PR.  Issues #409, #378, #358 will address this.
